src/query/ee/src/storages/fuse/operations/vacuum_table_v2.rs

```diff
@@ -0,0 +1,320 @@
+async fn set_lvt(fuse_table: &FuseTable, ctx: &dyn TableContext) -> Result<Option<DateTime<Utc>>> {
+    let Some(latest_snapshot) = fuse_table.read_table_snapshot().await? else {
+        info!(
+            "Table {} has no snapshot, stop vacuuming",
+            fuse_table.get_table_info().desc
+        );
+        return Ok(None);
+    };
+    if !is_uuid_v7(&latest_snapshot.snapshot_id) {
+        info!(
+            "latest snapshot {:?} is not v7, stop vacuuming",
+            latest_snapshot
+        );
+        return Ok(None);
+    }
+    // safe to unwrap, as we have checked the version is v5
+    let latest_ts = latest_snapshot.timestamp.unwrap();
+    let lvt_point_candidate =
+        Utc::now() - Days::new(ctx.get_settings().get_data_retention_time_in_days()?);
+    let lvt_point_candidate = std::cmp::min(lvt_point_candidate, latest_ts);
+    let cat = ctx.get_default_catalog()?;
+    let lvt_point = cat
+        .set_table_lvt(SetLVTReq {
+            table_id: fuse_table.get_table_info().ident.table_id,
+            time: lvt_point_candidate,
+        })
+        .await?
+        .time;
+    Ok(Some(lvt_point))
+}
+
+fn is_uuid_v7(uuid: &Uuid) -> bool {
+    let version = uuid.get_version();
+    version.is_some_and(|v| matches!(v, Version::SortRand))
+}
+
+async fn list_until_prefix(
+    fuse_table: &FuseTable,
+    path: &str,
+    until: &str,
+    need_one_more: bool,
+) -> Result<Vec<String>> {
+    let mut lister = fuse_table.get_operator().lister(path).await?;
+    let mut paths = vec![];
+    while let Some(entry) = lister.try_next().await? {
+        if entry.path() >= until {
+            if need_one_more {
+                paths.push(entry.path().to_string());
+            }
+            break;
+        }
+        paths.push(entry.path().to_string());
+    }
+    Ok(paths)
+}
+
+async fn list_until_timestamp(
+    fuse_table: &FuseTable,
+    path: &str,
+    until: DateTime<Utc>,
+    need_one_more: bool,
+) -> Result<Vec<String>> {
+    let uuid = uuid_from_date_time(until);
+    let uuid_str = uuid.simple().to_string();
+
+    // extract the most significant 48 bits, which is 12 characters
+    let timestamp_component = &uuid_str[..12];
+    let until = format!("{}g{}", path, timestamp_component);
+    list_until_prefix(fuse_table, path, &until, need_one_more).await
+}
+
+async fn read_snapshot_from_location(
+    fuse_table: &FuseTable,
+    path: &str,
+) -> Result<Arc<TableSnapshot>> {
+    let reader = MetaReaders::table_snapshot_reader(fuse_table.get_operator());
+    let ver = TableMetaLocationGenerator::snapshot_version(path);
+    let params = LoadParams {
+        location: path.to_owned(),
+        len_hint: None,
+        ver,
+        put_cache: false,
+    };
+    reader.read(&params).await
+}
+
+async fn select_gc_root<'a>(
+    fuse_table: &FuseTable,
+    snapshots_before_lvt: &'a [String],
+) -> Result<Option<(Arc<TableSnapshot>, &'a [String])>> {
```

Comment:
if session.retention_period .. is 0, we can/should use the current snapshot as GC root